src/targets/gpu/kernels/include/migraphx/kernels/scatter_reduction_modes.hpp

```diff
@@ -21,31 +21,63 @@
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  * THE SOFTWARE.
  */
-#ifndef MIGRAPHX_GUARD_OPERATORS_SCATTER_MUL_HPP
-#define MIGRAPHX_GUARD_OPERATORS_SCATTER_MUL_HPP
+#ifndef MIGRAPHX_GUARD_KERNELS_SCATTER_REDUCTION_MODES_HPP
+#define MIGRAPHX_GUARD_KERNELS_SCATTER_REDUCTION_MODES_HPP
 
-#include <array>
-#include <migraphx/config.hpp>
-#include <migraphx/op/scatter.hpp>
+#include <migraphx/kernels/types.hpp>
 
-// Scatter op. with "multiply" as the reduction function.
 namespace migraphx {
-inline namespace MIGRAPHX_INLINE_NS {
-namespace op {
 
-struct scatter_mul : scatter<scatter_mul>
+struct assign_none
 {
-    // reduction (pointwise operation) is called by the parent struct's compute() method.
-    // It works much like a virtual function overload.
-    // For the scatter operators, there are three different reduction functions.
-    auto reduction() const
+    template <class T, class U>
+    MIGRAPHX_DEVICE_CONSTEXPR void operator()(T& x, U y) const
     {
-        return [](auto& x, const auto& y) { x *= y; };
+        x = y;
     }
 };
 
-} // namespace op
-} // namespace MIGRAPHX_INLINE_NS
-} // namespace migraphx
+struct assign_add
+{
+    template <class T, class U>
+    MIGRAPHX_DEVICE_CONSTEXPR void operator()(T& x, U y) const
+    {
+        atomicAdd(&x, y);
+    }
+};
 
+struct assign_mul
```

Comment:
Use `op::product` instead of this class.

Reply:
op::product does not perform an atomic operation, which would lead to incorrect results.